### DISCUSSION TITLE
How to get regex submatches? (using `--json`)

### DISCUSSION THREAD

**Discussion by bew** [BurntSushi/ripgrep#2814] [Q&A] [answered]
Hello!

I have a number of files with lines like `type: foo, bar` or `type: baz`.
I'm looking to get the `foo, bar` & `baz` texts out of the JSON output.

Repro:
```console
$ cd `mktemp -d`
$ echo 'type: foo, bar' > file1
$ echo 'type: baz' > file2

$ rg --json '^type: (.*)' | rg '"type":"match"'
2:{"type":"match","data":{"path":{"text":"file2"},"lines":{"text":"type: baz\n"},"line_number":1,"absolute_offset":0,"submatches":[{"match":{"text":"type: baz"},"start":0,"end":9}]}}
5:{"type":"match","data":{"path":{"text":"file1"},"lines":{"text":"type: foo, bar\n"},"line_number":1,"absolute_offset":0,"submatches":[{"match":{"text":"type: foo, bar"},"start":0,"end":14}]}}
```
We can see the submatches:
```json
[{"match":{"text":"type: baz"},"start":0,"end":9}]
```
```json
[{"match":{"text":"type: foo, bar"},"start":0,"end":14}]
```
But I'd expect to get (for each):
- submatch 0, for the whole match
- submatch 1, with only the text after `type: ` so `foo, bar` for the file1 & `baz` for file2.

I also tried with `--regexp` without change.
Versions tried: `13.0.0` & `14.1.0`

**Comment by BurntSushi** [comment]
Could you please provide a full reproduction? Give me a command that I can run as-is. You've _described_ your input, but you need to _show_ your input. It's okay to make something up that models your real data, just so long as it exhibits the same problem.

**Reply by bew** [reply]
You're totally right, sorry about that 😬. I updated my post with setup commands and updated outputs.

**Comment by BurntSushi** [comment, marked as answer]
OK, so the issue here is that `submatches` is not the same as capture group matches. `submatches` in this context refers to [the list of matches within a line](https://docs.rs/grep-printer/latest/grep_printer/struct.JSON.html#message-match). For example:

```
$ echo 'foo bar baz' | rg --json '\w+' | rg '"match"'
{"type":"match","data":{"path":{"text":"<stdin>"},"lines":{"text":"foo bar baz\n"},"line_number":1,"absolute_offset":0,"submatches":[{"match":{"text":"foo"},"start":0,"end":3},{"match":{"text":"bar"},"start":4,"end":7},{"match":{"text":"baz"},"start":8,"end":11}]}}
```

Otherwise, I think you're probably looking for https://github.com/BurntSushi/ripgrep/issues/2325. The main issue is that adding capture groups will result in a perf hit. So it means that you have to opt into it. Which is kinda unfortunate.

**Reply by bew** [reply]
Oh I see the confusion, thank you I'll subscribe to that issue in case it's ever going to be implemented.
I can match the groups on my side in nushell so it's not a priority for me.

**Comment by Aleem28** [comment]
hello mate .. iam new in github and i have no idea of how to use these files. I want to download some files in rapidgator but the time and speed restriction prevent me to do so
can you do a video on youtube explaing step by step how to install rapidgator downloder. How to use python script. I found few links for rapidgator but all files are in .py format I don't know how to use or run. Can anyone explain please.

**Reply by BurntSushi** [reply]
Sorry, but no, I'm not going to make a YouTube video for you. I think you'll have better luck if you seek out a newbie forum. Not sure where though. Good luck.
